xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,62 +274,92 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+    int ret = -1;
     int call_count = 0;
+    int i = (long)cookie;
+    int need_open_count = 0;
 
     priv = this->private;
+    local->replies[i].valid = 1;
+    local->replies[i].op_ret = op_ret;
+    local->replies[i].op_errno = op_errno;
+    if (op_ret != 0) {
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, AFR_MSG_LK_HEAL_DOM,
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    } else {
+        local->cont.lk.getlk_rsp1[i] = *lock;
+    }
 
-    if (!afr_is_fd_fixable(fd))
-        goto out;
+    if (!local->strict_locks) {
+        ret = dict_get_int32(xdata, "strict-locks", &local->strict_locks);
+        if (ret)
+            gf_msg(this->name, GF_LOG_ERROR, -ret, AFR_MSG_DICT_GET_FAILED,
+                   "Failed to get value of strict-locks setting");
+    }
 
-    fd_ctx = afr_fd_ctx_get(fd, this);
-    if (!fd_ctx)
-        goto out;
+    call_count = afr_frame_return(frame);
 
-    need_open = alloca0(priv->child_count);
+    if (call_count)
+        return 0;
 
-    call_count = afr_fd_ctx_need_open(fd, this, need_open);
-    if (!call_count)
-        goto out;
+    if (!local->strict_locks) {
+        goto reopen_fd;
+    }
 
-    frame = create_frame(this, this->ctx->pool);
-    if (!frame)
-        goto out;
+    for (i = 0; i < priv->child_count; i++) {
+        if (!local->replies[i].valid || local->replies[i].op_ret != 0) {
+            continue;
+        }
+        if (local->cont.lk.getlk_rsp1[i].l_type == F_UNLCK) {
+            continue;
+        } else if (local->cont.lk.getlk_rsp1[i].l_type == F_WRLCK) {
+            locks_held = _gf_true;
+            break;
+        }
+    }
 
-    local = AFR_FRAME_INIT(frame, op_errno);
-    if (!local)
+    success_replies = alloca0(priv->child_count);
+    afr_fill_success_replies(local, priv, success_replies);
+    if (AFR_COUNT(success_replies, priv->child_count) == 0) {
         goto out;
+    }
 
-    local->loc.inode = inode_ref(fd->inode);
-    ret = loc_path(&local->loc, NULL);
-    if (ret < 0)
+    if (locks_held) {
+        gf_log(this->name, GF_LOG_INFO,
+               "client.strict-locks is enabled and conflicting locks are held. "
+               "Skipping fd reopen");
         goto out;
+    }
 
-    local->fd = fd_ref(fd);
-    local->fd_ctx = fd_ctx;
-
-    local->call_count = call_count;
+reopen_fd:
+    need_open = alloca0(priv->child_count);
 
-    gf_msg_debug(this->name, 0, "need open count: %d", call_count);
+    need_open_count = afr_fd_ctx_need_open(local->fd, this, need_open);
+    call_count = need_open_count;
+    if (!need_open_count) {
+        goto out;
+    }
+    gf_msg_debug(this->name, 0, "need open count: %d", need_open_count);
+    local->call_count = need_open_count;
 
     for (i = 0; i < priv->child_count; i++) {
         if (!need_open[i])
             continue;
 
-        if (IA_IFDIR == fd->inode->ia_type) {
+        local->fd_ctx->opened_on[i] = AFR_FD_OPENING;
```

Comment:
This is moved out of afr_fd_ctx_need_open, since we call that function from multiple places now, and we need to set this only when we proceed with the fix_open. In other case we should not set this.
Moved this inside a LOCK now.